enhancements/baremetal/baremetal-config-raid-and-bios.md

```diff
@@ -0,0 +1,151 @@
+- IPI deployments can handle the configuration of RAID and BIOS.
+
+### Non-Goals
+
+- Add new or vendor specific options regarding RAID and BIOS.
+
+## Proposal
+
+The configuration of RAID and BIOS is finally handed over to Ironic during the IPI deployments.
+1. The user runs `create manifests` and then adds RAID and BIOS fields to the **BMH** created by the installer.
+2. The interface to terraform in the installer code parses the BIOS and RAID fields.
+3. [Terraform-provider-ironic](https://github.com/openshift-metal3/terraform-provider-ironic) gets the
+configuration of RAID and BIOS for control plane hosts and then passes it to Ironic([Baremetal Operator(BMO)](https://github.com/metal3-io/baremetal-operator)
+has supported the process of RAID and BIOS for worker nodes).
+
+### User Stories
+
+With the addition of this feature, the users can configure RAID and BIOS in IPI deployments.
+
+### Implementation Details/Notes/Constraints
+
+#### Add two fields
+
+The user runs `create manifests` and then adds RAID and BIOS fields to the **BMH** created by the installer.
+
+```yaml
+spec:
+  bmc:
+    address: <bmc>://<ip-address>
+    credentialsName: worker-0-bmc-secret
+  bootMACAddress: mac-address
+  raid: raid-config
+  firmware: bios-config
+  ...
+```
+
+RAID feature has been implemented in **BMO**, so the *raid* field here is the same as the [BMH](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/config/crd/bases/metal3.io_baremetalhosts.yaml).
+The *firmware* field is the same as the *spec.firmware* field in **BMH** which is been advancing by [#302](https://github.com/metal3-io/baremetal-operator/pull/302).
+
+#### Process the fields in installer
+
+For control plane hosts, transform the *raid* and *firmware* fields into json respectively and then write them into the
+***terraform.baremetal.auto.tfvars.json*** file as the values of *raid_config* and *bios_settings* respectively.
+
+#### Process the fields in terraform-provider-ironic
+
+Add *raid_config* and *bios_settings* fields to terraform-provider-ironic API, transform them back to struct format, and then call the
+methods in **BMO** to process the two fields to perform [manual cleaning](https://docs.openstack.org/ironic/latest/admin/cleaning.html#manual-cleaning).
+
+#### Notes
+
+When the IPI deployments completed, the **BMH** resources will still get persisted in the cluster, though control plane hosts have
+RAID and BIOS field in their **BMH**, but the *externallyProvisioned* field's value is true, so the RAID and BIOS configurations are
+invalid.
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Test Plan
+
+- Unit tests for determining the configuration of RAID and BIOS passed to Ironic meeting expectations.
+- e2e tests for determining the configuration of RAID and BIOS configured during the IPI deployments.
+
+### Graduation Criteria
+
+TBD
```

Comment:
You need to fill all `TBD` with something, at least with `N/A` if the item does not apply.

Reply:
Done.